### DISCUSSION TITLE
How to use `notifyUpdates()` and `markTablesUpdated()` and what is the difference between them?

### DISCUSSION THREAD

**Discussion by ivanburlakov** [simolus3/drift#1384] [Q&A] [answered]
I searched the whole internet and can't seem to find an answer on how to use `notifyUpdates()` and `markTablesUpdated()`.
When I do it like this: `Provider.of<AppDatabase>(context, listen: false).markTablesUpdated({Stopwatches});` (`Stopwatches` is my table class), I get this error: `The element type 'Type' can't be assigned to the set type 'TableInfo<Table, dynamic>'.` I would appreciate any help on how to get this to work. And would also like to know what is the difference between those methods. Thank you! 😊

**Comment by simolus3** [comment, marked as answer]
`markTablesUpdated` marks a table as updated (duh :D). This means that all active streams which might depend on that table will be getting updated. It takes a set of `TableInfo` objects. `TableInfo` is the base class for generated moor tables. You can obtain them using the database class:

```dart
final db = Provider.of<AppDatabase>(context, listen: false);
db.markTablesUpdated({db.stopwatches});
```

`notifyUpdates` is more precise: Instead of just describing the tables that changed, it can describe what kind of update has happened (insert, update or delete). This is used for stream updates in complex scenarios. For instance, consider 

- Two tables, `a` and `b`
- A [trigger](https://sqlite.org/lang_createtrigger.html) inserting a row into `b` after a delete on `a`
- A stream listening on `b`

If we obverse a delete on table `a`, the stream needs to be updated since the trigger might have changed table `b . On the other hand, we don't need to update the stream after an insert on `a` because it only affects unrelated tables. So it may be desirable to have more information available on table updates in some cases.

**Reply by ivanburlakov** [reply]
Thank you so much for the detailed response, I got all to work perfectly 😊 I think this needs to be in the docs, you can edit my question or transfer it to issues accordingly (I don't know how to do this right 😚).

Dear @simolus3 , thank you so much for your timely answers and the maintance of this repository. I am using this library and browsing the repo for about 2 days, but I already got the feeling, that this is one of the nicest communities of any subject I ever witnessed. Thank you so much one more time 😉
